beacon-chain/blockchain/service.go

```diff
@@ -275,80 +252,87 @@ func (c *ChainService) processBlock(block *types.Block) error {
 		return fmt.Errorf("block points to nil parent: %#x", block.ParentHash())
 	}
 
-	aState, err := c.beaconDB.GetActiveState()
+	beaconState, err := c.beaconDB.GetState()
 	if err != nil {
-		return fmt.Errorf("failed to get active state: %v", err)
-	}
-	cState, err := c.beaconDB.GetCrystallizedState()
-	if err != nil {
-		return fmt.Errorf("failed to get crystallized state: %v", err)
+		return fmt.Errorf("failed to get beacon state: %v", err)
 	}
 
-	if valid := block.IsValid(
-		c.beaconDB,
-		aState,
-		cState,
+	if valid := state.IsValidBlock(
+		block,
+		beaconState,
 		parent.SlotNumber(),
 		c.genesisTime,
 	); !valid {
 		return errors.New("block failed validity conditions")
 	}
 
-	if err = c.calculateNewBlockVotes(block, aState, cState); err != nil {
-		return fmt.Errorf("failed to update block vote cache: %v", err)
+	if err := c.calculateNewBlockVotes(block, beaconState); err != nil {
+		return fmt.Errorf("failed to calculate block vote cache: %v", err)
 	}
 
 	// First, include new attestations to the active state
 	// so that they're accounted for during cycle transitions.
-	aState = aState.UpdateAttestations(block.Attestations())
+	beaconState.SetPendingAttestations(block.Attestations())
 
 	// If the block is valid, we compute its associated state tuple (active, crystallized)
-	var didCycleTransition bool
-	if cState.IsCycleTransition(block.SlotNumber()) {
-		cState, err = c.executeStateTransition(cState, aState, block)
+	if beaconState.IsCycleTransition(block.SlotNumber()) {
+		beaconState, err = c.executeStateTransition(beaconState, block, parent.SlotNumber())
```

Comment:
I find it fishy that this state transition function is calculating the next state by accepting the block of that next state as a parameter. This may be partially contributing to the circular reference problem between blocks and states. This sorta works in this situation since the next block is given to us, but will be an awkward interface when a local validator wants to generate the next block.